File: test/scripts/06-false-sharing/correct.js
/**
 * CORRECT: Each worker has local counter (no shared hot data)
 */
const { Worker } = require('worker_threads');
const path = require('path');
const NUM_WORKERS = 4;
const DURATION_SEC = parseInt(process.env.DURATION_SEC || '300', 10);
const stopSab = new SharedArrayBuffer(4);
const stopFlag = new Int32Array(stopSab);
const workerPath = path.join(__dirname, '_worker-local.js');
const workers = [];
for (let i = 0; i < NUM_WORKERS; i++) {
  workers.push(new Promise((r) => {
    const w = new Worker(workerPath, { workerData: { stopSab } });
    w.on('exit', r);
  }));
}
console.log('CORRECT: Per-worker local counter');
setTimeout(() => { Atomics.store(stopFlag, 0, 1); Promise.all(workers).then(() => console.log('Stopped')); }, DURATION_SEC * 1000);
process.on('SIGINT', () => Atomics.store(stopFlag, 0, 1));

File: test/scripts/06-false-sharing/correct.py
#!/usr/bin/env python3
"""
CORRECT: Separate counters per process (no shared array) - no false sharing.
Resume: "Avoid False Sharing" - Per-thread data.
"""
import multiprocessing as mp
import os
import signal
import time
NUM_PROCS = 4
ITERATIONS = 500_000
DURATION_SEC = int(os.environ.get("DURATION_SEC", "300"))
stop = mp.Value('i', 0)

def worker(index):
    local_count = 0
    while not stop.value:
        for _ in range(ITERATIONS):
            local_count += 1
    return local_count

def main():
    print("CORRECT: Per-process local counter -> no false sharing")
    signal.signal(signal.SIGINT, lambda s, f: setattr(stop, 'value', 1))
    signal.signal(signal.SIGTERM, lambda s, f: setattr(stop, 'value', 1))
    procs = [mp.Process(target=worker, args=(i,)) for i in range(NUM_PROCS)]
    for p in procs:
        p.start()
    time.sleep(DURATION_SEC)
    stop.value = 1
    for p in procs:
        p.join(timeout=2)
    print("Stopped")

if __name__ == "__main__":
    main()
